File: internal/models/service_analytics.go
package models

import (
	"agent-desk/internal/pkg/enums"
	"time"
)

// ConversationServiceSession is the rebuildable operational fact for one
// conversation session. It never participates in routing or reply decisions.
type ConversationServiceSession struct {
	ID                    int64                           `gorm:"primaryKey;autoIncrement"`
	TenantID              int64                           `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_service_session,priority:1"`
	ConversationID        int64                           `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_service_session,priority:2"`
	SessionNo             int                             `gorm:"type:int;not null;default:1;index;uniqueIndex:uk_service_session,priority:3"`
	CustomerID            int64                           `gorm:"type:bigint;not null;default:0;index"`
	ChannelID             int64                           `gorm:"type:bigint;not null;default:0;index"`
	StoreID               int64                           `gorm:"type:bigint;not null;default:0;index"`
	WxWorkInstanceID      int64                           `gorm:"type:bigint;not null;default:0;index"`
	ServiceMode           enums.IMConversationServiceMode `gorm:"type:int;not null;default:3;index"`
	Status                enums.ServiceSessionStatus      `gorm:"type:varchar(20);not null;default:'open';index"`
	StartedAt             time.Time                       `gorm:"type:datetime;not null;index"`
	QueueEnteredAt        *time.Time                      `gorm:"type:datetime;index"`
	AssignedAt            *time.Time                      `gorm:"type:datetime;index"`
	FirstHumanReplyAt     *time.Time                      `gorm:"type:datetime;index"`
	LastHumanReplyAt      *time.Time                      `gorm:"type:datetime;index"`
	EndedAt               *time.Time                      `gorm:"type:datetime;index"`
	FirstAssignmentID     int64                           `gorm:"type:bigint;not null;default:0;index"`
	LastAssignmentID      int64                           `gorm:"type:bigint;not null;default:0;index"`
	AssignedTeamID        int64                           `gorm:"type:bigint;not null;default:0;index"`
	AssignedSquadID       int64                           `gorm:"type:bigint;not null;default:0;index"`
	AssignedAgentID       int64                           `gorm:"type:bigint;not null;default:0;index"`
	CustomerMessageCount  int                             `gorm:"type:int;not null;default:0"`
	AIMessageCount        int                             `gorm:"type:int;not null;default:0"`
	HumanMessageCount     int                             `gorm:"type:int;not null;default:0"`
	SystemMessageCount    int                             `gorm:"type:int;not null;default:0"`
	AssignmentCount       int                             `gorm:"type:int;not null;default:0"`
	TransferCount         int                             `gorm:"type:int;not null;default:0"`
	HumanHandled          bool                            `gorm:"not null;default:false;index"`
	AIHandled             bool                            `gorm:"not null;default:false;index"`
	QueueSeconds          int64                           `gorm:"type:bigint;not null;default:0"`
	FirstResponseSeconds  int64                           `gorm:"type:bigint;not null;default:0"`
	TotalHumanWaitSeconds int64                           `gorm:"type:bigint;not null;default:0"`
	LastMessageID         int64                           `gorm:"type:bigint;not null;default:0;index"`
	LastMessageAt         *time.Time                      `gorm:"type:datetime;index"`
	CloseReason           string                          `gorm:"type:varchar(255);not null;default:''"`
	ResolutionCode        string                          `gorm:"type:varchar(50);not null;default:'';index"`
	CategoryCode          string                          `gorm:"type:varchar(50);not null;default:'';index"`
	TagIDsJSON            string                          `gorm:"type:text"`
	SessionSummary        string                          `gorm:"type:text"`
	FactOrigin            enums.AnalyticsFactOrigin       `gorm:"type:varchar(20);not null;default:'runtime';index"`
	DataQuality           enums.AnalyticsDataQuality      `gorm:"type:varchar(20);not null;default:'exact';index"`
	EstimatedFieldsJSON   string                          `gorm:"type:text"`
	AuditFields
}

// ConversationResponseSpan measures one customer message burst until the next
// human agent reply. AI replies never close a span.
type ConversationResponseSpan struct {
	ID                     int64                      `gorm:"primaryKey;autoIncrement"`
	TenantID               int64                      `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_response_span_start,priority:1"`
	ConversationID         int64                      `gorm:"type:bigint;not null;default:0;index"`
	SessionNo              int                        `gorm:"type:int;not null;default:1;index"`
	AssignmentID           int64                      `gorm:"type:bigint;not null;default:0;index"`
	TeamID                 int64                      `gorm:"type:bigint;not null;default:0;index"`
	SquadID                int64                      `gorm:"type:bigint;not null;default:0;index"`
	AgentID                int64                      `gorm:"type:bigint;not null;default:0;index"`
	CustomerStartMessageID int64                      `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_response_span_start,priority:2"`
	CustomerEndMessageID   int64                      `gorm:"type:bigint;not null;default:0;index"`
	CustomerMessageCount   int                        `gorm:"type:int;not null;default:1"`
	StartedAt              time.Time                  `gorm:"type:datetime;not null;index"`
	RepliedAt              *time.Time                 `gorm:"type:datetime;index"`
	ReplyMessageID         int64                      `gorm:"type:bigint;not null;default:0;index"`
	WaitSeconds            int64                      `gorm:"type:bigint;not null;default:0"`
	Status                 enums.ResponseSpanStatus   `gorm:"type:varchar(20);not null;default:'waiting';index"`
	FactOrigin             enums.AnalyticsFactOrigin  `gorm:"type:varchar(20);not null;default:'runtime';index"`
	DataQuality            enums.AnalyticsDataQuality `gorm:"type:varchar(20);not null;default:'exact';index"`
	EstimatedFieldsJSON    string                     `gorm:"type:text"`
	AuditFields
}

type AgentPresenceSession struct {
	ID              int64                     `gorm:"primaryKey;autoIncrement"`
	TenantID        int64                     `gorm:"type:bigint;not null;default:0;index"`
	UserID          int64                     `gorm:"type:bigint;not null;default:0;index"`
	AgentProfileID  int64                     `gorm:"type:bigint;not null;default:0;index"`
	TeamID          int64                     `gorm:"type:bigint;not null;default:0;index"`
	Status          enums.AgentPresenceStatus `gorm:"type:varchar(20);not null;default:'online';index"`
	Source          string                    `gorm:"type:varchar(40);not null;default:'';index"`
	BreakReason     string                    `gorm:"type:varchar(100);not null;default:'';index"`
	ChangedBy       int64                     `gorm:"type:bigint;not null;default:0;index"`
	StartedAt       time.Time                 `gorm:"type:datetime;not null;index"`
	LastSeenAt      time.Time                 `gorm:"type:datetime;not null;index"`
	EndedAt         *time.Time                `gorm:"type:datetime;index"`
	DurationSeconds int64                     `gorm:"type:bigint;not null;default:0"`
	AuditFields
}

type QualityTemplate struct {
	ID          int64        `gorm:"primaryKey;autoIncrement"`
	TenantID    int64        `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_quality_template_version,priority:1"`
	Name        string       `gorm:"type:varchar(120);not null;default:'';index;uniqueIndex:uk_quality_template_version,priority:2"`
	Description string       `gorm:"type:text"`
	TotalScore  int          `gorm:"type:int;not null;default:100"`
	PassScore   int          `gorm:"type:int;not null;default:80"`
	Version     int          `gorm:"type:int;not null;default:1;uniqueIndex:uk_quality_template_version,priority:3"`
	IsDefault   bool         `gorm:"not null;default:false;index"`
	Status      enums.Status `gorm:"type:int;not null;default:0;index"`
	AuditFields
}

type QualityTemplateItem struct {
	ID          int64                 `gorm:"primaryKey;autoIncrement"`
	TenantID    int64                 `gorm:"type:bigint;not null;default:0;index"`
	TemplateID  int64                 `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_quality_template_item,priority:1"`
	Code        string                `gorm:"type:varchar(64);not null;default:'';index;uniqueIndex:uk_quality_template_item,priority:2"`
	Name        string                `gorm:"type:varchar(120);not null;default:''"`
	Description string                `gorm:"type:text"`
	RuleType    enums.QualityRuleType `gorm:"type:varchar(20);not null;default:'score';index"`
	MetricCode  string                `gorm:"type:varchar(64);not null;default:'';index"`
	MaxScore    int                   `gorm:"type:int;not null;default:0"`
	Required    bool                  `gorm:"not null;default:true"`
	HardFail    bool                  `gorm:"not null;default:false"`
	SortNo      int                   `gorm:"type:int;not null;default:0;index"`
	Status      enums.Status          `gorm:"type:int;not null;default:0;index"`
	AuditFields
}

// QualityInspection scores only the human replies written during one
// ConversationAssignment segment.
type QualityInspection struct {
	ID             int64                         `gorm:"primaryKey;autoIncrement"`
	TenantID       int64                         `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_quality_assignment,priority:1"`
	ConversationID int64                         `gorm:"type:bigint;not null;default:0;index"`
	SessionNo      int                           `gorm:"type:int;not null;default:1;index"`
	AssignmentID   int64                         `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_quality_assignment,priority:2"`
	AgentID        int64                         `gorm:"type:bigint;not null;default:0;index"`
	TeamID         int64                         `gorm:"type:bigint;not null;default:0;index"`
	TemplateID     int64                         `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_quality_assignment,priority:3"`
	Status         enums.QualityInspectionStatus `gorm:"type:varchar(20);not null;default:'draft';index"`
	TotalScore     int                           `gorm:"type:int;not null;default:0"`
	MaxScore       int                           `gorm:"type:int;not null;default:100"`
	HardFailed     bool                          `gorm:"not null;default:false;index"`
	Result         enums.QualityInspectionResult `gorm:"type:varchar(20);not null;default:'';index"`
	Summary        string                        `gorm:"type:text"`
	InspectedBy    int64                         `gorm:"type:bigint;not null;default:0;index"`
	InspectedAt    *time.Time                    `gorm:"type:datetime;index"`
	AuditFields
}

type QualityInspectionItem struct {
	ID             int64                 `gorm:"primaryKey;autoIncrement"`
	TenantID       int64                 `gorm:"type:bigint;not null;default:0;index"`
	InspectionID   int64                 `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_quality_inspection_item,priority:1"`
	TemplateItemID int64                 `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_quality_inspection_item,priority:2"`
	ItemCode       string                `gorm:"type:varchar(64);not null;default:'';index"`
	ItemName       string                `gorm:"type:varchar(120);not null;default:''"`
	RuleType       enums.QualityRuleType `gorm:"type:varchar(20);not null;default:'score';index"`
	MaxScore       int                   `gorm:"type:int;not null;default:0"`
	Score          int                   `gorm:"type:int;not null;default:0"`
	Passed         bool                  `gorm:"not null;default:false"`
	HardFailed     bool                  `gorm:"not null;default:false"`
	MetricValue    string                `gorm:"type:varchar(100);not null;default:''"`
	Evidence       string                `gorm:"type:text"`
	MessageIDsJSON string                `gorm:"type:text"`
	Comment        string                `gorm:"type:text"`
	AuditFields
}

type QualitySamplingBatch struct {
	ID           int64                       `gorm:"primaryKey;autoIncrement"`
	TenantID     int64                       `gorm:"type:bigint;not null;default:0;index"`
	Name         string                      `gorm:"type:varchar(120);not null;default:'';index"`
	CriteriaJSON string                      `gorm:"type:text"`
	Seed         string                      `gorm:"type:varchar(64);not null;default:'';index"`
	SampleSize   int                         `gorm:"type:int;not null;default:0"`
	Status       enums.QualitySamplingStatus `gorm:"type:varchar(20);not null;default:'ready';index"`
	CreatedBy    int64                       `gorm:"type:bigint;not null;default:0;index"`
	CompletedAt  *time.Time                  `gorm:"type:datetime;index"`
	AuditFields
}

type QualitySamplingItem struct {
	ID             int64 `gorm:"primaryKey;autoIncrement"`
	TenantID       int64 `gorm:"type:bigint;not null;default:0;index"`
	BatchID        int64 `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_quality_sample_item,priority:1"`
	AssignmentID   int64 `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_quality_sample_item,priority:2"`
	ConversationID int64 `gorm:"type:bigint;not null;default:0;index"`
	SessionNo      int   `gorm:"type:int;not null;default:1;index"`
	AgentID        int64 `gorm:"type:bigint;not null;default:0;index"`
	InspectionID   int64 `gorm:"type:bigint;not null;default:0;index"`
	AuditFields
}

type DispatchDecisionLog struct {
	ID                    int64                        `gorm:"primaryKey;autoIncrement"`
	TenantID              int64                        `gorm:"type:bigint;not null;default:0;index"`
	DecisionKey           string                       `gorm:"type:varchar(128);not null;default:'';uniqueIndex"`
	ConversationID        int64                        `gorm:"type:bigint;not null;default:0;index"`
	SessionNo             int                          `gorm:"type:int;not null;default:1;index"`
	AssignmentID          int64                        `gorm:"type:bigint;not null;default:0;index"`
	Trigger               string                       `gorm:"type:varchar(40);not null;default:'';index"`
	DecisionMode          string                       `gorm:"type:varchar(30);not null;default:'';index"`
	Status                enums.DispatchDecisionStatus `gorm:"type:varchar(20);not null;default:'';index"`
	CandidateUserIDsJSON  string                       `gorm:"type:text"`
	CandidateSnapshotJSON string                       `gorm:"type:text"`
	InputLastMessageID    int64                        `gorm:"type:bigint;not null;default:0;index"`
	SelectedUserID        int64                        `gorm:"type:bigint;not null;default:0;index"`
	SelectedTeamID        int64                        `gorm:"type:bigint;not null;default:0;index"`
	SelectedSquadID       int64                        `gorm:"type:bigint;not null;default:0;index"`
	DecisionLatencyMillis int64                        `gorm:"type:bigint;not null;default:0"`
	Reason                string                       `gorm:"type:text"`
	FallbackReason        string                       `gorm:"type:text"`
	OperatorID            int64                        `gorm:"type:bigint;not null;default:0;index"`
	DecidedAt             time.Time                    `gorm:"type:datetime;not null;index"`
	AuditFields
}

type ServiceAnalyticsPolicy struct {
	ID                         int64 `gorm:"primaryKey;autoIncrement"`
	TenantID                   int64 `gorm:"type:bigint;not null;default:0;uniqueIndex"`
	QueueTargetSeconds         int   `gorm:"type:int;not null;default:60"`
	FirstResponseTargetSeconds int   `gorm:"type:int;not null;default:180"`
	ResponseTargetSeconds      int   `gorm:"type:int;not null;default:300"`
	RepeatConsultationHours    int   `gorm:"type:int;not null;default:24"`
	SatisfactionThreshold      int   `gorm:"type:int;not null;default:4"`
	EvaluationExpiryHours      int   `gorm:"type:int;not null;default:72"`
	DefaultSampleSize          int   `gorm:"type:int;not null;default:20"`
	AuditFields
}

type ConversationEvaluation struct {
	ID             int64                              `gorm:"primaryKey;autoIncrement"`
	TenantID       int64                              `gorm:"type:bigint;not null;default:0;index"`
	ConversationID int64                              `gorm:"type:bigint;not null;default:0;index"`
	SessionNo      int                                `gorm:"type:int;not null;default:1;index"`
	AssignmentID   int64                              `gorm:"type:bigint;not null;default:0;index"`
	CustomerID     int64                              `gorm:"type:bigint;not null;default:0;index"`
	Status         enums.ConversationEvaluationStatus `gorm:"type:varchar(20);not null;default:'pending';index"`
	InviteChannel  string                             `gorm:"type:varchar(30);not null;default:'';index"`
	TokenHash      string                             `gorm:"type:varchar(64);not null;default:'';uniqueIndex"`
	InvitedBy      int64                              `gorm:"type:bigint;not null;default:0;index"`
	InvitedAt      time.Time                          `gorm:"type:datetime;not null;index"`
	ExpiresAt      time.Time                          `gorm:"type:datetime;not null;index"`
	SubmittedAt    *time.Time                         `gorm:"type:datetime;index"`
	Rating         int                                `gorm:"type:int;not null;default:0;index"`
	TagCodesJSON   string                             `gorm:"type:text"`
	Comment        string                             `gorm:"type:text"`
	AuditFields
}

type ReportViewPreset struct {
	ID          int64        `gorm:"primaryKey;autoIncrement"`
	TenantID    int64        `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_report_view_preset,priority:1"`
	UserID      int64        `gorm:"type:bigint;not null;default:0;index;uniqueIndex:uk_report_view_preset,priority:2"`
	PageCode    string       `gorm:"type:varchar(50);not null;default:'';index;uniqueIndex:uk_report_view_preset,priority:3"`
	Name        string       `gorm:"type:varchar(100);not null;default:'';uniqueIndex:uk_report_view_preset,priority:4"`
	FiltersJSON string       `gorm:"type:text"`
	ColumnsJSON string       `gorm:"type:text"`
	SortJSON    string       `gorm:"type:text"`
	IsDefault   bool         `gorm:"not null;default:false;index"`
	Status      enums.Status `gorm:"type:int;not null;default:0;index"`
	AuditFields
}

File: web/lib/api/service-analytics.ts
import { request, requestBlob } from "@/lib/api/client"
import {
  AgentPresenceStatus,
  AnalyticsDataQuality,
  AnalyticsFactOrigin,
  ConversationEvaluationStatus,
  QualityInspectionStatus,
  QualityRuleType,
  QualitySamplingStatus,
} from "@/lib/generated/enums"

export type Paging = { page: number; limit: number; total: number }
export type PageResult<T> = { results: T[]; page: Paging }

function queryString(values: Record<string, string | number | boolean | undefined>) {
  const query = new URLSearchParams()
  Object.entries(values).forEach(([key, value]) => {
    if (value === undefined || value === "") return
    query.set(key, String(value))
  })
  const output = query.toString()
  return output ? `?${output}` : ""
}

export type AnalyticsSummary = {
  sessionCount: number
  uniqueCustomerCount: number
  closedSessionCount: number
  humanQueueCount: number
  assignedCount: number
  humanRepliedCount: number
  unansweredCount: number
  queueFailureCount: number
  transferSessionCount: number
  repeatConsultationCount: number
  totalMessageCount: number
  customerMessageCount: number
  aiMessageCount: number
  humanMessageCount: number
  assignmentAccessRate: number
  effectiveAccessRate: number
  transferRate: number
  repeatConsultationRate: number
  averageQueueSeconds: number
  p50QueueSeconds: number
  p90QueueSeconds: number
  averageFirstReplySeconds: number
  p50FirstReplySeconds: number
  p90FirstReplySeconds: number
  averageResponseSeconds: number
  p50ResponseSeconds: number
  p90ResponseSeconds: number
  averageHumanWaitSeconds: number
  p50HumanWaitSeconds: number
  p90HumanWaitSeconds: number
  averageSessionSeconds: number
  p50SessionSeconds: number
  p90SessionSeconds: number
  averageMessagesPerSession: number
  queueSlaRate: number
  firstReplySlaRate: number
  responseSlaRate: number
  qualityInspectableCount: number
  qualityInspectionCount: number
  qualityPendingCount: number
  qualityPassedCount: number
  qualityFailedCount: number
  qualityCoverageRate: number
  qualityPassRate: number
  averageQualityScore: number
  evaluationInviteCount: number
  evaluationSubmittedCount: number
  satisfiedCount: number
  evaluationParticipationRate: number
  satisfactionRate: number
  averageSatisfaction: number
  exactSessionCount: number
  estimatedSessionCount: number
  incompleteSessionCount: number
}

export type AnalyticsTrend = {
  date: string
  sessions: number
  humanQueues: number
  humanReplies: number
  messages: number
  averageQueue: number
  averageFirstReply: number
  averageResponse: number
  averageSession: number
}

export type AnalyticsAgent = {
  agentId: number
  agentName: string
  teamId: number
  teamName: string
  squadNames: string[]
  currentStatus: "offline" | AgentPresenceStatus | string
  currentActiveCount: number
  maxConcurrentCount: number
  assignedCount: number
  repliedCount: number
  unansweredCount: number
  humanMessageCount: number
  responseCount: number
  serviceSeconds: number
  averageFirstReplySeconds: number
  p50FirstReplySeconds: number
  p90FirstReplySeconds: number
  averageResponseSeconds: number
  p50ResponseSeconds: number
  p90ResponseSeconds: number
  responseSlaRate: number
  onlineSeconds: number
  idleSeconds: number
  busySeconds: number
  breakSeconds: number
  firstOnlineAt?: string
  lastOnlineAt?: string
  utilizationRate: number
  qualityInspectableCount: number
  qualityInspectionCount: number
  qualityPendingCount: number
  qualityPassedCount: number
  qualityFailedCount: number
  qualityPassRate: number
  averageQualityScore: number
  evaluationInviteCount: number
  evaluationSubmittedCount: number
  satisfiedCount: number
  evaluationParticipationRate: number
  satisfactionRate: number
  averageSatisfaction: number
}

export type AnalyticsSource = {
  storeId: number
  storeName: string
  wxWorkInstanceId: number
  wxWorkEmployeeName: string
  sessionCount: number
  humanQueueCount: number
  humanRepliedCount: number
  messageCount: number
  averageFirstReply: number
  effectiveAccessRate: number
  qualityInspectableCount: number
  qualityInspectionCount: number
  qualityPassedCount: number
  qualityCoverageRate: number
  qualityPassRate: number
  averageQualityScore: number
  evaluationInviteCount: number
  evaluationSubmittedCount: number
  satisfiedCount: number
  evaluationParticipationRate: number
  satisfactionRate: number
  averageSatisfaction: number
}

export type AnalyticsRealtime = {
  openSessionCount: number
  aiActiveCount: number
  queueingCount: number
  assignedActiveCount: number
  waitingReplyCount: number
  longestQueueSeconds: number
  queueSlaAlertCount: number
  onlineAgentCount: number
  idleAgentCount: number
  busyAgentCount: number
  breakAgentCount: number
  offlineAgentCount: number
  availableCapacity: number
  todaySessionCount: number
  todayQueueCount: number
  todayAssignedCount: number
  todayHumanRepliedCount: number
  todayTransferCount: number
  todayQueueFailureCount: number
  todayMessageCount: number
  todayAverageQueueSeconds: number
  todayAverageFirstReplySeconds: number
}

export type AnalyticsDistribution = {
  key: string
  label: string
  count: number
  rate: number
}

export type AnalyticsOverview = {
  startAt: string
  endAt: string
  generatedAt: string
  summary: AnalyticsSummary
  realtime: AnalyticsRealtime
  trend: AnalyticsTrend[]
  firstReplyDistribution: AnalyticsDistribution[]
  responseDistribution: AnalyticsDistribution[]
  sessionDurationDistribution: AnalyticsDistribution[]
  agents: AnalyticsAgent[]
  sources: AnalyticsSource[]
  dispatch: {
    decisionCount: number
    selectedCount: number
    autoCount: number
    manualCount: number
    ruleCount: number
    modelCount: number
    hybridCount: number
    fallbackCount: number
    failedCount: number
    staleCount: number
    overrideCount: number
    transferCount: number
    autoRate: number
    averageDecisionLatencyMillis: number
  }
}

export type AnalyticsDimensionItem = {
  id: number
  name: string
  parentId: number
}

export type AnalyticsDimensions = {
  teams: AnalyticsDimensionItem[]
  squads: AnalyticsDimensionItem[]
  agents: AnalyticsDimensionItem[]
  channels: AnalyticsDimensionItem[]
  stores: AnalyticsDimensionItem[]
  wxWorkInstances: AnalyticsDimensionItem[]
}

export type ServiceSession = {
  id: number
  conversationId: number
  sessionNo: number
  customerId: number
  customerName: string
  channelId: number
  channelName: string
  storeId: number
  storeName: string
  wxWorkInstanceId: number
  wxWorkEmployeeName: string
  status: string
  startedAt: string
  queueEnteredAt?: string
  assignedAt?: string
  firstHumanReplyAt?: string
  endedAt?: string
  assignedTeamId: number
  assignedTeamName: string
  assignedAgentId: number
  assignedAgentName: string
  customerMessageCount: number
  aiMessageCount: number
  humanMessageCount: number
  assignmentCount: number
  transferCount: number
  queueSeconds: number
  firstResponseSeconds: number
  totalHumanWaitSeconds: number
  closeReason: string
  lastMessageAt?: string
  resolutionCode: string
  categoryCode: string
  tagIds: number[]
  sessionSummary: string
  factOrigin: AnalyticsFactOrigin
  dataQuality: AnalyticsDataQuality
  estimatedFields: string[]
}

export type SessionMessage = {
  id: number
  conversationId: number
  senderType: "customer" | "agent" | "ai" | "system" | string
  senderId: number
  senderName?: string
  messageType: string
  content: string
  payload?: string
  seqNo: number
  sentAt?: string
  recalledAt?: string
}

export type QualityTemplateItem = {
  id: number
  code: string
  name: string
  description: string
  ruleType: QualityRuleType
  metricCode: string
  maxScore: number
  required: boolean
  hardFail: boolean
  sortNo: number
}

export type QualityTemplate = {
  id: number
  name: string
  description: string
  totalScore: number
  passScore: number
  version: number
  isDefault: boolean
  items: QualityTemplateItem[]
}

export type QualityInspectionItem = {
  templateItemId: number
  itemCode: string
  itemName: string
  ruleType: QualityRuleType
  maxScore: number
  score: number
  passed: boolean
  hardFailed: boolean
  metricValue: string
  evidence: string
  messageIds: number[]
  comment: string
}

export type QualityInspection = {
  id: number
  conversationId: number
  sessionNo: number
  assignmentId: number
  agentId: number
  agentName: string
  teamId: number
  teamName: string
  templateId: number
  status: QualityInspectionStatus
  totalScore: number
  maxScore: number
  result: string
  hardFailed: boolean
  summary: string
  inspectedBy: number
  inspectedAt?: string
  items: QualityInspectionItem[]
}

export type QualityPoolEntry = {
  assignmentId: number
  conversationId: number
  sessionNo: number
  customerName: string
  agentId: number
  agentName: string
  teamId: number
  teamName: string
  storeName: string
  wxWorkEmployeeName: string
  assignedAt: string
  finishedAt?: string
  humanReplyCount: number
  inspection?: QualityInspection
}

export function fetchAnalyticsOverview(query: Record<string, string | number | undefined>) {
  return request<AnalyticsOverview>(`/api/dashboard/service-analytics/overview${queryString(query)}`)
}

export function exportAnalyticsOverview(query: Record<string, string | number | undefined>) {
  return requestBlob(`/api/dashboard/service-analytics/export${queryString(query)}`)
}

export function fetchAnalyticsDimensions() {
  return request<AnalyticsDimensions>("/api/dashboard/service-analytics/dimensions")
}

export function fetchServiceSessions(query: Record<string, string | number | boolean | undefined>) {
  return request<PageResult<ServiceSession>>(`/api/dashboard/service-session/list${queryString(query)}`)
}

export function fetchServiceSessionDimensions() {
  return request<AnalyticsDimensions>("/api/dashboard/service-session/dimensions")
}

export function fetchServiceSessionMessages(sessionId: number, page = 1, limit = 100) {
  return request<PageResult<SessionMessage>>(
    `/api/dashboard/service-session/message_list${queryString({ sessionId, page, limit })}`,
  )
}

export function fetchServiceSession(sessionId: number) {
  return request<ServiceSession>(`/api/dashboard/service-session/${sessionId}`)
}

export function updateServiceSessionAnnotation(payload: {
  id: number
  resolutionCode: string
  categoryCode: string
  sessionSummary: string
  tagIds: number[]
}) {
  return request<ServiceSession>("/api/dashboard/service-session/annotate", {
    method: "POST",
    body: JSON.stringify(payload),
  })
}

export function serviceSessionExportPath(
  query: Record<string, string | number | boolean | undefined>,
) {
  return `/api/dashboard/service-session/export${queryString(query)}`
}

export function exportServiceSessions(
  query: Record<string, string | number | boolean | undefined>,
) {
  return requestBlob(serviceSessionExportPath(query))
}

export function fetchQualityPool(query: Record<string, string | number | undefined>) {
  return request<PageResult<QualityPoolEntry>>(`/api/dashboard/quality-inspection/pool${queryString(query)}`)
}

export function fetchQualityTemplates() {
  return request<QualityTemplate[]>("/api/dashboard/quality-template/list")
}

export function saveQualityTemplate(payload: {
  id?: number
  name: string
  description: string
  passScore: number
  isDefault: boolean
  items: Array<{
    id?: number
    code: string
    name: string
    description: string
    ruleType: QualityRuleType
    metricCode: string
    maxScore: number
    required: boolean
    hardFail: boolean
    sortNo: number
  }>
}) {
  return request<QualityTemplate>("/api/dashboard/quality-template/save", {
    method: "POST",
    body: JSON.stringify(payload),
  })
}

export function fetchQualityInspection(id: number) {
  return request<QualityInspection>(`/api/dashboard/quality-inspection/${id}`)
}

export function saveQualityInspection(payload: {
  id?: number
  assignmentId: number
  templateId: number
  status: QualityInspectionStatus
  summary: string
  items: Array<{
    templateItemId: number
    score: number
    violated?: boolean
    evidence: string
    messageIds: number[]
    comment: string
  }>
}) {
  return request<QualityInspection>("/api/dashboard/quality-inspection/save", {
    method: "POST",
    body: JSON.stringify(payload),
  })
}

export type ServiceAnalyticsPolicy = {
  queueTargetSeconds: number
  firstResponseTargetSeconds: number
  responseTargetSeconds: number
  repeatConsultationHours: number
  satisfactionThreshold: number
  evaluationExpiryHours: number
  defaultSampleSize: number
}

export function fetchAnalyticsPolicy() {
  return request<ServiceAnalyticsPolicy>("/api/dashboard/service-analytics/policy")
}

export function updateAnalyticsPolicy(payload: ServiceAnalyticsPolicy) {
  return request<ServiceAnalyticsPolicy>("/api/dashboard/service-analytics/policy/update", {
    method: "POST",
    body: JSON.stringify(payload),
  })
}

export type QualitySamplingItem = {
  assignmentId: number
  conversationId: number
  sessionNo: number
  agentId: number
  inspectionId: number
}

export type QualitySamplingBatch = {
  id: number
  name: string
  criteriaJson: string
  seed: string
  sampleSize: number
  status: QualitySamplingStatus
  createdBy: number
  createdAt: string
  completedAt?: string
  items: QualitySamplingItem[]
}

export function createQualitySampling(payload: {
  name: string
  teamId?: number
  agentId?: number
  startAt: string
  endAt: string
  sampleSize: number
}) {
  return request<QualitySamplingBatch>("/api/dashboard/quality-sampling/create", {
    method: "POST",
    body: JSON.stringify(payload),
  })
}

export function fetchQualitySampling(id: number) {
  return request<QualitySamplingBatch>(`/api/dashboard/quality-sampling/${id}`)
}

export function fetchQualitySamplingList(query: Record<string, string | number | undefined>) {
  return request<PageResult<QualitySamplingBatch>>(
    `/api/dashboard/quality-sampling/list${queryString(query)}`,
  )
}

export type ReportViewPreset = {
  id: number
  pageCode: string
  name: string
  filtersJson: string
  columnsJson: string
  sortJson: string
  isDefault: boolean
}

export function fetchReportViewPresets(pageCode: string) {
  return request<ReportViewPreset[]>(
    `/api/dashboard/report-view-preset/list${queryString({ pageCode })}`,
  )
}

export function saveReportViewPreset(payload: Omit<ReportViewPreset, "id"> & { id?: number }) {
  return request<ReportViewPreset>("/api/dashboard/report-view-preset/save", {
    method: "POST",
    body: JSON.stringify(payload),
  })
}

export function deleteReportViewPreset(id: number) {
  return request<void>(`/api/dashboard/report-view-preset/delete${queryString({ id })}`, {
    method: "POST",
  })
}

export type AgentPresence = {
  status: "offline" | AgentPresenceStatus
  breakReason: string
  startedAt?: string
  lastSeenAt?: string
}

export function fetchCurrentAgentPresence() {
  return request<AgentPresence>("/api/dashboard/agent-presence/current")
}

export function updateAgentPresence(payload: {
  status: AgentPresenceStatus
  breakReason?: string
}) {
  return request<AgentPresence>("/api/dashboard/agent-presence/update", {
    method: "POST",
    body: JSON.stringify(payload),
  })
}

export type ConversationEvaluation = {
  id: number
  conversationId: number
  sessionNo: number
  assignmentId: number
  customerId: number
  status: ConversationEvaluationStatus
  inviteChannel: string
  invitedAt: string
  expiresAt: string
  submittedAt?: string
  rating: number
  tagCodes: string[]
  comment: string
}

export type ConversationEvaluationInvite = {
  evaluation: ConversationEvaluation
  path: string
}

export function inviteConversationEvaluation(payload: {
  serviceSessionId: number
  assignmentId?: number
}) {
  return request<ConversationEvaluationInvite>("/api/dashboard/conversation-evaluation/invite", {
    method: "POST",
    body: JSON.stringify(payload),
  })
}

export function fetchConversationEvaluations(query: Record<string, string | number | undefined>) {
  return request<PageResult<ConversationEvaluation>>(
    `/api/dashboard/conversation-evaluation/list${queryString(query)}`,
  )
}

export type PublicConversationEvaluation = {
  status: ConversationEvaluationStatus
  companyName: string
  expiresAt: string
  submittedAt?: string
  rating: number
}

export function validateConversationEvaluation(token: string) {
  return request<PublicConversationEvaluation>(
    `/api/evaluation/validate${queryString({ token })}`,
    { skipAuth: true },
  )
}

export function submitConversationEvaluation(payload: {
  token: string
  rating: number
  tagCodes: string[]
  comment: string
}) {
  return request<PublicConversationEvaluation>("/api/evaluation/submit", {
    method: "POST",
    body: JSON.stringify(payload),
    skipAuth: true,
  })
}
